docs/DEVELOPER_GUIDE.md

```diff
@@ -0,0 +1,258 @@
+- If needed, create new `conftest.py` files in the relevant directories.
+
+
+## Fixtures
+- Ordering: Always call pytest native fixtures first, then session-scoped fixtures and then any other fixtures.
+- Fixtures should handle setup (and the teardown, if needed) needed for the test(s), including the creation of resources, for example.
+- Fixtures should do one thing only.
+For example, instead of:
+
+```python
+@pytest.fixture()
+def network_vm():
+    with NetworkAttachmentDefinition(name=...) as nad:
+      with VirtualMachine(name=..) as vm:
+        yield vm
+```
+
+Do:
+
+```python
+@pytest.fixture()
+def network_attachment_definition():
+    with NetworkAttachmentDefinition(name=...) as nad:
+      yield nad
+
+@pytest.fixture(network_attachment_definition)
+def model_inference_service(network_attachment_definition):
+    with VirtualMachine(name=..) as vm:
+        yield vm
+
+```
+
+- Pytest reports failures in fixtures as ERROR
+- A fixture name should be a noun that describes what the fixture provides (i.e. returns or yields), rather than a verb.
+For example:
+  - If a test needs a storage secret, the fixture should be called 'storage_secret' and not 'create_secret'.
+  - If a test needs a directory to store user data, the fixture should be called 'user_data_dir' and not 'create_directory'.
+- Note fixture scope, test execution times can be reduced by selecting the right scope.
+Pytest default fixture invocation is "function", meaning the code in the fixture will be executed every time the fixture is called.
+Broader scopes (class, module etc) will invoke the code only once within the given scope and all tests within the scope will use the same instance.
+- Use request.param to pass parameters from test/s to fixtures; use a dict structure for readability.  For example:
+
+```code
+@pytest.mark.parametrize(
+"my_secret",
+[
+pytest.param(
+{"name": "my-secret", "data-dict": {"key": "value"}}},
+),
+
+def test_secret(my_secret):
+
+    pass
+
+@pytest.fixture()
+def my_secret(request):
+secret = Secret(name=request.param["name"], data_dict=request.param["data-dict"])
+```
+
+
+## Tests
+- Pytest reports failures in fixtures as FAILED
+- Each test should have a clear purpose and should be easy to understand.
+- Each test should verify a single aspect of the product.
+- Preferably, each test should be independent of other tests.
+- When there's a dependency between tests use pytest dependency plugin to mark the relevant hierarchy between tests (https://github.com/RKrahl/pytest-dependency)
+- When adding a new test, apply relevant marker(s) which may apply.
+Check [pytest.ini](../pytest.ini) for available markers; additional markers can always be added when needed.
+- Classes are good to group related tests together, for example, when they share a fixture.
+You should NOT group unrelated tests in one class (because it is misleading the reader).
+
+
+# Development
```

Comment:
I guess this is the setup env.
I think this needs to come before coding guidelines.

Thinking about it, perhaps coding standards and guiding should be extracted to a different file.